eCommerce Sample Data

alexfrancoeur opened the pull request:
### Screenshots

<img width="1440" alt="screen shot 2018-09-14 at 11 35 07 pm" src="https://user-images.githubusercontent.com/25181823/45582030-f7551480-b876-11e8-9030-d89cba2d20bd.png">

<img width="1440" alt="screen shot 2018-09-14 at 11 35 37 pm" src="https://user-images.githubusercontent.com/25181823/45582032-fc19c880-b876-11e8-801a-1b15737db643.png">

_Edit: Dashboard screenshot updated on Sep 18_
![image](https://user-images.githubusercontent.com/25181823/45694107-88a0e100-bb2c-11e8-97b5-89801fd8f82d.png)

### Sample document
```json
{
  "_index": "kibana_sample_data_ecommerce",
  "_type": "_doc",
  "_id": "EZ2y2mUBi4dtX2qiTQME",
  "_version": 1,
  "_score": null,
  "_source": {
    "category": [
      "Men's Shoes",
      "Men's Clothing"
    ],
    "currency": "EUR",
    "customer_birth_date": "1995-03-05",
    "customer_first_name": "Youssef",
    "customer_full_name": "Youssef Alvarez",
    "customer_gender": "MALE",
    "customer_id": 31,
    "customer_last_name": "Alvarez",
    "customer_phone": "",
    "day_of_week": "Saturday",
    "day_of_week_i": 5,
    "email": "youssef@alvarez-family.zzz",
    "manufacturer": [
      "Pier One"
    ],
    "order_date": "2017-09-02T23:45:36+00:00",
    "order_id": 592109,
    "products": [
      {
        "base_price": 49.99,
        "discount_percentage": 0,
        "quantity": 1,
        "manufacturer": "Pier One",
        "tax_amount": 0,
        "product_id": 12202,
        "category": "Men's Shoes",
        "sku": "ZO0243602436",
        "taxless_price": 49.99,
        "unit_discount_amount": 0,
        "min_price": 26.49,
        "_id": "sold_product_592109_12202",
        "discount_amount": 0,
        "created_on": "2016-12-31T23:45:36+00:00",
        "product_name": "Moccasins - stone",
        "price": 49.99,
        "taxful_price": 49.99,
        "base_unit_price": 49.99
      },
      {
        "base_price": 28.99,
        "discount_percentage": 0,
        "quantity": 1,
        "manufacturer": "Pier One",
        "tax_amount": 0,
        "product_id": 15017,
        "category": "Men's Clothing",
        "sku": "ZO0281402814",
        "taxless_price": 28.99,
        "unit_discount_amount": 0,
        "min_price": 13.63,
        "_id": "sold_product_592109_15017",
        "discount_amount": 0,
        "created_on": "2016-12-31T23:45:36+00:00",
        "product_name": "Jumper - off-white",
        "price": 28.99,
        "taxful_price": 28.99,
        "base_unit_price": 28.99
      }
    ],
    "sku": [
      "ZO0243602436",
      "ZO0281402814"
    ],
    "taxful_total_price": 78.98,
    "taxless_total_price": 78.98,
    "total_quantity": 2,
    "total_unique_products": 2,
    "type": "order",
    "user": "youssef",
    "geoip": {
      "country_iso_code": "US",
      "location": {
        "lon": -74,
        "lat": 40.8
      },
      "region_name": "New York",
      "continent_name": "North America",
      "city_name": "New York"
    }
  },
  "fields": {
    "customer_birth_date": [
      "1995-03-05T00:00:00.000Z"
    ],
    "order_date": [
      "2017-09-02T23:45:36.000Z"
    ]
  },
  "sort": [
    1504395936000
  ]
}
```

@jamiesmith @asawariS @AlonaNadler would love some input on the dashboard.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2653/)

nreese commented on config/kibana.yml:
```diff
@@ -111,3 +111,4 @@
 
 # Specifies locale to be used for all localizable strings, dates and number formats.
 #i18n.locale: "en"
+xpack.reporting.capture.browser.type: "chromium"
```
Do not check local changes into kibana.yml

alexfrancoeur replied:
woops, 🤕

nreese commented on src/server/sample_data/data_sets/ecommerce/index.js:
```diff
@@ -0,0 +1,191 @@
+      }
+    },
+    timeFields: ['order_date'],
+    currentTimeMarker: '2018-01-09T00:00:00',
```
`currentTimeMarker ` is causing the time issue you are seeing. `currentTimeMarker` is the reference to now in your data set.  When the data is installed, timestamps are translated based on the present time.  The distance between a timestamp and `currentTimeMarker` is used to translate the data set. For the eCommerce data, it looks like the values occur during December 2016. The reason you where seeing the big gap between the current date and the translated data is that your value of `2018-01-09T00:00:00` translated the data with a gap of 13 months.

<img width="2043" alt="screen shot 2018-09-17 at 5 47 13 am" src="https://user-images.githubusercontent.com/373691/45621489-b9561d00-ba3d-11e8-947d-ca9fda007a7b.png">

Setting this value to `2016-12-11T00:00:00` seems like a reasonable value that translates the data appropriately.

<img width="2036" alt="screen shot 2018-09-17 at 5 51 38 am" src="https://user-images.githubusercontent.com/373691/45621682-6761c700-ba3e-11e8-9813-faef48bcaf99.png">

alexfrancoeur replied:
ahh, thanks Nathan! Makes sense, oversight on my end.

nreese commented on src/server/sample_data/routes/lib/load_data.test.js:
```diff
@@ -42,3 +42,15 @@ test('load log data', done => {
+  const bulkInsertMock = (docs) => {
+    myDocsCount += docs.length;
+  };
+  loadData('./src/server/sample_data/data_sets/logs/logs.json.gz', bulkInsertMock, async (err, count) => {
```
The test is failing because the test is loading the `logs` data set.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2732/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2770/)

alexfrancoeur commented:
Fixed some mappings and added two more visualizations

![image](https://user-images.githubusercontent.com/25181823/45694107-88a0e100-bb2c-11e8-97b5-89801fd8f82d.png)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2819/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2829/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2859/)

nreese commented:
@alexfrancoeur I just merged a PR that will require you to update this PR. PR https://github.com/elastic/kibana/pull/23230 changed the sample data set schema to support multiple indices. You will need to update your schema provider to the new format which looks like the example below. I have also been putting the field mappings into their own file to improve readability. 

```
// new schema
{
    id: 'ecommerce',
    name: 'Sample eCommerce orders',
    description: 'Sample data, visualizations, and dashboards for tracking eCommerce orders.',
    previewImagePath: '/plugins/kibana/home/sample_data_resources/ecommerce/dashboard.png',
    overviewDashboard: '722b74f0-b882-11e8-a6d9-e546fe2bba5f',
    defaultIndex: 'ff959d40-b880-11e8-a6d9-e546fe2bba5f',
    savedObjects: savedObjects,
    dataIndices: [
        {
            id: 'ecommerce',
            dataPath: path.join(__dirname, './ecommerce.json.gz'),
            fields: fieldMappings,
            timeFields: ['order_date'],
            currentTimeMarker: '2016-12-11T00:00:00',
            preserveDayOfWeekTimeOfDay: true,
        }
    ]
}
```

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2918/)

alexfrancoeur commented:
@nreese think I got it right, mind taking a look?

nreese commented on src/server/sample_data/routes/lib/load_data.test.js:
```diff
@@ -38,3 +38,23 @@ test('load log data', async () => {
+  const bulkInsertMock = (docs) => {
+    myDocsCount += docs.length;
+  };
+<<<<<<< HEAD
```
There are still merge conflicts in the test file that need to be fixed

alexfrancoeur replied:
I believe I fixed this as well

nreese commented on test/functional/apps/home/_sample_data.js:
```diff
@@ -114,6 +127,26 @@ export default function ({ getService, getPageObjects }) {
+        const toTime = `${todayYearMonthDay} 23:59:59.999`;
+        await PageObjects.header.setAbsoluteRange(fromTime, toTime);
+        const panelCount = await PageObjects.dashboard.getPanelCount();
+<<<<<<< HEAD
```
merge conflicts here too

alexfrancoeur replied:
thinks this is good to go

nreese commented:
All of the code looks good now. Just wait for a green CI

Did you check your saved objects to make sure they were not effected by the rebase?

alexfrancoeur commented:
@nreese yep just ran locally, seems to have all the most recent changes. Thanks! Once I get some feedback from @AlonaNadler and @asawariS, I'll push that green button and backport

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3033/)

AlonaNadler commented:
@alexfrancoeur 

Nice job on e-commerce!!! I might forget how to ingest data since I keep using sample data 😆 
Few suggestions: 
- I didn't understand the `ring the bell` annotation, it made me go in and look, might be good it made me curious 
- Can you add an example for a scripted field, like if (category:women* AND customer_gender:Male) then field: gift (it's an assumption.... whatever works )
- how about using a workpad instead of (or in addition to)  a dashboard for this sample data
- In terms of the data maybe do one big spike or a drop just to spice things up (and maybe allow to use it with ML)
- The colors in this chart are not friendly, I missed the fact there are 4 series, also trousers seem to be on zero all the time 
![image](https://user-images.githubusercontent.com/32550316/45907712-ded08700-bdad-11e8-96bf-c5114c6ba285.png)

- Maybe provide an example with Timelion, if users use that as reference it might be beneficial

alexfrancoeur commented:
Thanks for the review @AlonaNadler!

@alexfrancoeur 

> - I didn't understand the `ring the bell` annotation, it made me go in and look, might be good it made me curious 

I could change the text here, but if made you curious enough to look I think I may leave it 😄 

> - Can you add an example for a scripted field, like if (category:women* AND customer_gender:Male) then field: gift (it's an assumption.... whatever works )

I really like the scripted field idea, especially the example you provided. I'm going to leave this as a future update though. Unfortunately, I'm just running out of time to get this in 🙁 

> - how about using a workpad instead of (or in addition to)  a dashboard for this sample data

Already on it with https://github.com/elastic/kibana/issues/22893

> - In terms of the data maybe do one big spike or a drop just to spice things up (and maybe allow to use it with ML)

Given the time to FF and priorities in the next coming weeks, I don't think I'll have time to adjust the data in this fashion. It would tie nicely with this issue though https://github.com/elastic/kibana/issues/18808

> - The colors in this chart are not friendly, I missed the fact there are 4 series, also trousers seem to be on zero all the time 

Ah! I borrowed the colors from an SA demo specifically. Trousers being 0 was due to a misspelling, fixed it. I adjusted the opacity and rearranged the series so it looks a bit better. If we need to adjust further, we can open a separate PR.

<img width="1494" alt="screen shot 2018-10-01 at 11 17 23 am" src="https://user-images.githubusercontent.com/25181823/46297822-9eba9280-c56b-11e8-8885-bf6fc67cd482.png">
 
> - Maybe provide an example with Timelion, if users use that as reference it might be beneficial

I like the idea and I think it'd be worth revisiting when it's not so close to FF. I'm going to prioritize getting this in vs. adding new types of visualizations. I think it'd be good to add an example in other sample data sets as well.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3752/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3773/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3793/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3866/)

alexfrancoeur commented:
jenkins, test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3920/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3943/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4026/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4071/)
